dask/dataframe/io/parquet/arrow.py

```diff
@@ -265,6 +87,426 @@ def _index_in_schema(index, schema):
+        (
+            index_names,
+            column_names,
+            storage_name_mapping,
+            column_index_names,
+        ) = _parse_pandas_metadata(pandas_metadata)
+        if categories is None:
+            categories = []
+            for col in pandas_metadata["columns"]:
+                if (col["pandas_type"] == "categorical") and (
+                    col["name"] not in categories
+                ):
+                    categories.append(col["name"])
+    else:
+        index_names = []
+        column_names = schema.names
+        storage_name_mapping = {k: k for k in column_names}
+        column_index_names = [None]
+
+    if index is None and index_names:
+        index = index_names
+
+    if set(column_names).intersection(partitions):
+        raise ValueError(
+            "partition(s) should not exist in columns.\n"
+            "categories: {} | partitions: {}".format(column_names, partitions)
+        )
+
+    column_names, index_names = _normalize_index_columns(
+        columns, column_names + partitions, index, index_names
+    )
+
+    all_columns = index_names + column_names
+
+    # Check that categories are included in columns
+    if categories and not set(categories).intersection(all_columns):
+        raise ValueError(
+            "categories not in available columns.\n"
+            "categories: {} | columns: {}".format(categories, list(all_columns))
+        )
+
+    dtypes = _get_pyarrow_dtypes(schema, categories)
+    dtypes = {storage_name_mapping.get(k, k): v for k, v in dtypes.items()}
+
+    index_cols = index or ()
+    meta = _meta_from_dtypes(all_columns, dtypes, index_cols, column_index_names)
+    meta = clear_known_categories(meta, cols=categories)
+
+    if partition_obj:
+        for partition in partition_obj:
+            if isinstance(index, list) and partition.name == index[0]:
+                meta.index = pd.CategoricalIndex(
+                    categories=partition.keys, name=index[0]
+                )
+            elif partition.name == meta.index.name:
+                meta.index = pd.CategoricalIndex(
+                    categories=partition.keys, name=meta.index.name
+                )
+            elif partition.name in meta.columns:
+                meta[partition.name] = pd.Series(
+                    pd.Categorical(categories=partition.keys, values=[]),
+                    index=meta.index,
+                )
+
+    return meta, index_cols, categories, index
+
+
+def _aggregate_stats(
+    file_path, file_row_group_stats, file_row_group_column_stats, stat_col_indices
+):
+    """ Utility to aggregate the statistics for N row-groups
+        into a single dictionary.
+    """
+    if len(file_row_group_stats) < 1:
+        raise ValueError("Empty statistics.")
+    elif len(file_row_group_column_stats) == 0:
+        assert len(file_row_group_stats) == 1
+        return file_row_group_stats[0]
+    else:
+        df_rgs = pd.DataFrame(file_row_group_stats)
```

Comment:
Right - I want to avoid python iteration when aggregating statistics over many row-groups (the above approach is fast in that case), but I'm alse sure there is a better way in general.

Reply:
Agreed, creating dataframes is slow (even compared to a dict of numpy arrays), but maybe this won't show up in profiling.

Reply:
I added some logic to avoid creating these dataframes in certain cases and added a note that we *should* try to avoid it altogether.  At the moment, doing this is worthwhile when we need to aggregate statistics for **many** row-groups, but is probably not worthwhile when aggregating a small collection of row-groups

Reply:
(if the number is small, it may indeed be wasteful per row group, but probably insignificant overall)